performance_manager/lib/rt_vehicle_positions.py

```diff
@@ -0,0 +1,277 @@
+import logging
+
+import pandas
+import numpy
+import sqlalchemy
+import re
+import datetime
+
+from sqlalchemy import select
+from sqlalchemy import update
+from sqlalchemy import insert
+from sqlalchemy import delete
+from sqlalchemy.orm import sessionmaker
+
+from typing import Optional
+from typing import Union
+from typing import List
+
+from .s3_utils import read_parquet
+from .postgres_utils import VehiclePositionEvents
+
+
+def get_vp_dataframe(to_load: Union[str, List[str]]) -> pandas.DataFrame:
+    """
+    return a dataframe from a vehicle position parquet file (or list of files)
+    with expected columns
+    """
+    vehicle_position_cols = [
+        "current_status",
+        "current_stop_sequence",
+        "stop_id",
+        "vehicle_timestamp",
+        "direction_id",
+        "route_id",
+        "start_date",
+        "start_time",
+        "vehicle_id",
+    ]
+
+    return read_parquet(to_load).loc[:, vehicle_position_cols]
+
+
+def start_time_to_seconds(
+    time: Optional[str],
+) -> Optional[float]:
+    """
+    transform time string in HH:MM:SS format to seconds integer
+    """
+    if time is None:
+        return time
+    (hour, min, sec) = time.split(":")
+    return int(hour) * 3600 + int(min) * 60 + int(sec)
+
+
+def transform_vp_dtyes(df: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    ingest dataframe of vehicle position data from parquet file and transform
+    column datatypes
+    """
+    # current_staus: 1 = MOVING, 0 = STOPPED_AT
+    df["current_status"] = numpy.where(
+        df["current_status"] != "STOPPED_AT", 1, 0
+    ).astype(numpy.byte)
+
+    # store start_date as Int64 [nullable] instead of string
+    df["start_date"] = pandas.to_numeric(df["start_date"]).astype("Int64")
+
+    # store current_stop_sequence as Int64 [nullable]
+    df["current_stop_sequence"] = pandas.to_numeric(
+        df["current_stop_sequence"]
+    ).astype("Int64")
+
+    # store direction_id as Int64 [nullable]
+    df["direction_id"] = pandas.to_numeric(df["direction_id"]).astype("Int64")
+
+    # store start_time as seconds from start of day (Int64 [nullable])
+    df["start_time"] = (
+        df["start_time"].apply(start_time_to_seconds).astype("Int64")
+    )
+
+    # hash row data, exluding vehicle_timestamp column
+    # This hash is used to identify continuous series of records with the same
+    # categorical data
+    df["hash"] = df[list(set(df.columns) - {"vehicle_timestamp"})].apply(
+        lambda row: hash(tuple(row)), axis=1
+    )
+
+    # drop duplicates of hash and vehicle_timestamp columns (extraneous data)
+    df = df.drop_duplicates(subset=["vehicle_timestamp", "hash"])
+
+    return df
+
+
+def transform_vp_timestamps(df: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    convert raw vehicle position data with one timestamp field into
+    vehicle position events with timestamp_start and timestamp_end
+    """
+    df = df.sort_values(by=["vehicle_id", "vehicle_timestamp"])
```

Comment:
Convert dataframe with one timestamp column into a dataframe with 2 timestamp columns. Requires the dataframe to be sorted by `vehicle_id` and `vehicle_timestamp` all operations are array based, avoid any iterating.